plugins/embark-assistant/survey.cpp

```diff
@@ -908,6 +908,22 @@ void inline copy_incursion_values(embark_assist::defs::mid_level_tile_incursion_
 
 //=================================================================================
 
+void reset_mlt_inorganics(embark_assist::defs::mid_level_tiles &mlts) {
+    const uint16_t size = mlts[0][0].metals.size();
+    for (uint8_t i = 0; i < 16; i++) {
+        for (uint8_t k = 0; k < 16; k++) {
+            embark_assist::defs::mid_level_tile &mlt = mlts[i][k];
+            for (uint16_t l = 0; l < size; l++) {
+                mlt.metals[l] = false;
+                mlt.economics[l] = false;
+                mlt.minerals[l] = false;
+            }
```

Comment:
`.at()` also checks the validity of the provided index and raises an `out_of_range_` exception if the index is out of bounds; `[]` does not do this, which could explain some of the speedups you're seeing if you've replaced `at()` with `[]`. I had assumed part of the reason for using `at()` earlier was for safety, but it seems like that may be less important here.

You are making an assumption that all of these vectors match the size of `mlts[0][0].metals`, though, and losing the safety checks from earlier. I might suggest this instead - it should be pretty well-optimized, and it will always use the right size:
```suggestion
            std::fill(mlt.metals.begin(), mlt.metals.end(), false);
            std::fill(mlt.economics.begin(), mlt.economics.end(), false);
            std::fill(mlt.minerals.begin(), mlt.minerals.end(), false);
```

Reply:
First off :+1:  for taking the time to look into the PR in such detail.

I'd love to use `std::fill` in fact I did initially in the [feature branch](https://github.com/bseiller/dfhack/blob/c9fc554f9362cc8163ebfa3a8c2ec47309f317bc/plugins/embark-assistant/survey.cpp#L1512) with the big rewrite, but as I wrote in the commit message I actually did performance comparisons between `.at()` , `[]` and `std::fill`, even looking at `std::vector.assign`.
And shockingly enough and counterintuitively `std:fill` takes more than twice as long to process the data than `[]`, `std::vector.assign` around 3 times more, at least on Windows and for C++11 (or is it C++14?) which we use for dfhack.
Of course this is specific to my machine, but for 16.908.544 tiles (large world) and 256 inorganics (assignments to the vector where applicable ) the different variants perform as follows:
- `[]` takes around 8 seconds. 
- `.at` takes around 9 seconds. 
- `std::fill` takes around 17 seconds. 
- `std::vector.assign` takes between 23 and 24 seconds.
- `std::vector.resize` (current solution previous to this PR) takes about as long as `std::vector.assign`.
- the hack as described at [[3]](https://stackoverflow.com/a/64075271/167865) takes 0,5 seconds, but it very convoluted...

And that is in standalone code, mind you.
The reason for this behavior concerning  `std::fill` probably is that std::vector<bool> has been neglected in MS C++/MSVC, missing the needed template specialization for the template functions such as `std::fill` => [[1]](https://isocpp.org/blog/2012/11/on-vectorbool) + [[2]](https://stackoverflow.com/questions/12433154/are-stdfill-stdcopy-specialized-for-stdvectorbool)

The use and performance of `std::vector<bool>` in context of the inorganics has been on my mind again and again during my work at embark_assistant. I thought about replacing it with lists that only contain the indices/ids of the found inorganics, even about a custom data structure, that is both fast and has a small memory footprint just like `std::vector<bool>` but doesn't need to be iterated over completely for access and reset. I event looked into providing the needed template specialization, but I couldn't find a good starting point for that.
I haven't found the "perfect" solution so far for this, one that does not include a lot of rewriting.

If you still favor the use of `std::fill`, I'll change it. Happily so if I knew that the performance would improve when migrating to a more recent version of the C++ Standard in the future.
Also I wonder if `std::vector<bool>` and `std::fill` perform better on *nix with GCC or Clang - do you have any insight in this? 
Alternatively to using `std::fill` I could call `size()` for every vector and use `.at` instead of `[]`, as the vectors are initialized once with `embark_assist::survey::initiate` and meant to have all the same size - which of course isn't guaranteed as in the future someone could call `embark_assist::survey::survey_mid_level_tile` with uninitialized vectors. But in that case there would be an error further down in `embark_assist::survey::survey_mid_level_tile` anyway even if the use `std::fill` prevents that initially.

links for reference:
[1] https://isocpp.org/blog/2012/11/on-vectorbool
[2] https://stackoverflow.com/questions/12433154/are-stdfill-stdcopy-specialized-for-stdvectorbool
[3] https://stackoverflow.com/a/64075271/167865

Reply:
I haven't actually done any of my own profiling on it. I did a bit of research into it yesterday, and it looked like `[]`, `assign()`, and `fill()` had nearly identical performance under GCC with `-O3` (what our release builds use). There's a chance that it was for more "normal" vectors, i.e. not `vector<bool>`. In any case, you've clearly done more testing here, so I'm fine going with the current implementation.

Reply:
As I was curious myself I did a simple comparison with `std::vector<char>` not changing anything else in the benchmark setup:
- `[]` takes around 4 to 5 seconds
- `std::fill` takes around 0,4 to 0,6 seconds as does `std::fill_n`
- `std::vector.assign` takes around 0,7 seconds
- `std::memset` takes around 0,5 seconds - this one can't be used for `std::vector<bool>` at all as it doesn't change the 
So `std::vector<bool>` is really a very special case: highly optimized in regards to the memory footprint but with a significant overhead for bulk manipulation, at least in my version of MSVC.
The introduction and the notes at the end [here](https://en.cppreference.com/w/cpp/container/vector_bool) convey the idea of how special `std::vector<bool>` pretty well.

And just to make sure `std::vector<char>` isn't a special case itself as well I also checked with `std::vector<int>`:
- `[]` takes around 6 seconds
- `std::fill` takes around 4 seconds
- `std::fill_n` takes around 2 seconds
- `std::vector.assign` takes around 3 seconds
- `std::memset` takes around 1 second
So `std::vector<char>` might be another special case with tailored template specializations or there might just be CPU instructions that can process chars faster than ints, but the bulk operations (`std::fill`, `std::fill_n` and `std::memset`) are faster for `std::vector<int>` than looped `[]` nonetheless. 

By now I'm thinking that the easiest and cleanest solution for this might be to replace the 3 `std::vector<bool>` in [mid_level_tile](https://github.com/DFHack/dfhack/blob/60c6c2a05976edfab329ff3e2b0d3b33f7476432/plugins/embark-assistant/defs.h#L86) with `std::vector<char>` and use `std::fill` indead.
As PR #1757 reduced the pressure to keep the storage of inorganics occurrence as memory efficient as possible and there are only 2 instances of https://github.com/DFHack/dfhack/blob/60c6c2a05976edfab329ff3e2b0d3b33f7476432/plugins/embark-assistant/defs.h#L99 and thus 512 instances of mid_level_tile the additional memory consumption would be about 336kiB, assuming there are 256 non-generated inorganics. I think that is acceptable considering it will save another 7 seconds of search time.
@lethosor, @PatrikLundell: What do the both of you think?
I'll look into it, but that is definitely a new PR.

Reply:
I agree. Using a vector over char rather than bool isn't butt ugly, saves a bit of time, and costs insignificant amounts of memory if there are only two copies of the set. I'd make sure to make a comment to explain the rationale for usage of char rather than bool, though.

Reply:
Makes sense to me too, although I'd suggest `vector<int8_t>` or `vector<uint8_t>` for consistency with existing DF/DFHack code.

Reply:
Thanks for pointing out that convention considering `vector<int8_t>`/`vector<uint8_t>` - I haven't spent much time looking into other plugins or the DFHack core itself, so getting this kind of hint is much appreciated & I'll heed this advice.